sled-agent/config-reconciler/src/reconciler_task/zones.rs

```diff
@@ -0,0 +1,1119 @@
+                        }
+
+                        ZoneState::FailedToStart(err) => {
+                            info!(
+                                log,
+                                "starting shutdown of a failed-to-start zone";
+                                "zone" => z.config.zone_name(),
+                                "prev_err" => InlineErrorChain::new(err),
+                            );
+                            true
+                        }
+                    },
+                }
+            })
+            .peekable();
+
+        // Save a bit of work: bail out now if we have no zones to stop.
+        if zones_to_shut_down.peek().is_none() {
+            return Ok(());
+        }
+
+        // Map the zones to the futures that will try to shut them down.
+        let shutdown_futures = zones_to_shut_down.map(|zone| {
+            zone.try_shut_down(sled_agent_facilities, zone_facilities, log)
+                .map(|result| (zone.config.id, result))
+        });
+
+        // Concurrently stop the zones, then record the results.
+        let shutdown_results = future::join_all(shutdown_futures).await;
+
+        let mut nfailed = 0;
+        for (zone_id, result) in shutdown_results {
+            match result {
+                Ok(()) => {
+                    self.zones.remove(&zone_id);
+                }
+                Err((state, err)) => {
+                    nfailed += 1;
+                    self.zones
+                        .get_mut(&zone_id)
+                        .expect("shutdown task operates on existing zone")
+                        .state = ZoneState::PartiallyShutDown { state, err };
+                }
+            }
+        }
+
+        // Report how many shutdowns failed.
+        match NonZeroUsize::new(nfailed) {
+            None => Ok(()),
+            Some(nfailed) => Err(nfailed),
+        }
+    }
+
+    /// Attempt to start any zones that are present in `desired_zones` but not
+    /// in `self`.
+    pub(super) async fn start_zones_if_needed<T: SledAgentFacilities>(
+        &mut self,
+        desired_zones: &IdMap<OmicronZoneConfig>,
+        sled_agent_facilities: &T,
+        is_time_synchronized: bool,
+        all_u2_pools: &CurrentlyManagedZpools,
+        log: &Logger,
+    ) {
+        // Filter desired zones down to just those that we need to start
+        let mut zones_to_start = desired_zones
+            .iter()
+            .filter(|zone| {
+                match self.zones.get(&zone.id).map(|z| &z.state) {
+                    // This is entirely new zone - start it!
+                    None => {
+                        info!(
+                            log, "starting new zone";
+                            "config" => ?zone,
+                        );
+                        true
+                    }
+
+                    // This is a zone we've tried to start before; try again!
+                    Some(ZoneState::FailedToStart(err)) => {
```

Comment:
I was slightly confused by the difference between the behavior here and that in `PartiallyShutDown` below. However it actually makes sense because if we fail to shutdown a node in `FailedToStart` inside `shut_down_zones_if_needed_impl` then the state transitions to `ZoneState::PartiallyShutDown`. There's never an instance where we try to start a node, fail, try to shut it down, fail, and then try to start it again while partially failed, which alleviates my initial concern.

It may be worth adding a note about this behavior and also specifying somewhere that `start_zones_if_needed` and `shut_down_zones_if_needed` cannot be called concurrently. I think we'd run into issues if they do and I don't believe running concurrently is your intent.

Reply:
> It may be worth adding a note about this behavior

Yeah, can do; this is not the easiest thing to follow (and I'm still not 100% sure it's correct in all the ways we could fail).

> specifying somewhere that start_zones_if_needed and shut_down_zones_if_needed cannot be called concurrently

Both of these take `&mut self` so can't be called concurrently! Do you think it should still be noted explicitly?

Reply:
> Both of these take &mut self so can't be called concurrently! Do you think it should still be noted explicitly?

Oh, no that makes sense. No need to document explicitly.

Reply:
> > It may be worth adding a note about this behavior
> 
> Yeah, can do; this is not the easiest thing to follow (and I'm still not 100% sure it's correct in all the ways we could fail).

Added in [b2583a7](https://github.com/oxidecomputer/omicron/pull/8140/commits/b2583a79d18df2494c6c145c332b0b859e951205)